File: packages/core-ts/src/revocation.ts
import * as ed25519 from '@noble/ed25519';
import { readFile } from 'node:fs/promises';
import type { Keypair, RevocationStatement } from './types.js';
import { getAgentId } from './identity.js';
import { encodeBase64 } from './utils.js';
import { verifyRevocationSignature } from './verification.js';

/**
 * Canonical field order for revocation statement signing.
 * Ensures deterministic serialization regardless of object property order.
 */
function canonicalizeRevocationPayload(
  statement: Omit<RevocationStatement, 'signature'>
): string {
  return JSON.stringify({
    type: statement.type,
    agent_id: statement.agent_id,
    revoked_key_id: statement.revoked_key_id,
    revoked_at: statement.revoked_at,
    reason: statement.reason,
  });
}

/** Create a signed revocation statement for a compromised key. */
export async function createRevocation(
  agentId: string,
  revokedKeyId: string,
  reason: string,
  signingKeypair: Keypair
): Promise<RevocationStatement> {
  const statement: Omit<RevocationStatement, 'signature'> = {
    type: 'revocation',
    agent_id: agentId,
    revoked_key_id: revokedKeyId,
    revoked_at: new Date().toISOString(),
    reason,
  };

  const payload = canonicalizeRevocationPayload(statement);
  const payloadBytes = new TextEncoder().encode(payload);
  const signatureBytes = await ed25519.signAsync(payloadBytes, signingKeypair.privateKey);

  return {
    ...statement,
    signature: encodeBase64(signatureBytes),
  };
}

/** Check if a given key ID appears in a revocation list. */
export function checkRevocation(
  keyId: string,
  revocationList: RevocationStatement[]
): { revoked: boolean; statement?: RevocationStatement } {
  const match = revocationList.find(
    (r) => r.revoked_key_id === keyId || r.agent_id === keyId
  );
  return match ? { revoked: true, statement: match } : { revoked: false };
}

export interface LoadRevocationListOptions {
  /**
   * If true, verify each revocation statement's signature and discard invalid ones.
   * Defaults to true. Set to false only for debugging.
   */
  verifySignatures?: boolean;
}

/** Load a revocation list from a file path or URL, optionally verifying signatures. */
export async function loadRevocationList(
  source: string,
  options: LoadRevocationListOptions = {}
): Promise<RevocationStatement[]> {
  const verifySignatures = options.verifySignatures ?? true;
  let raw: string;

  if (source.startsWith('http://') || source.startsWith('https://')) {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 10_000);
    try {
      const response = await fetch(source, { signal: controller.signal });
      if (!response.ok) {
        throw new Error(`Failed to fetch revocation list from ${source}: ${response.status}`);
      }
      raw = await response.text();
      if (raw.length > 10 * 1024 * 1024) {
        throw new Error('Revocation list exceeds 10MB size limit');
      }
    } finally {
      clearTimeout(timeout);
    }
  } else {
    raw = await readFile(source, 'utf-8');
  }

  const parsed = JSON.parse(raw);
  if (!Array.isArray(parsed)) {
    throw new Error('Revocation list must be a JSON array');
  }

  const statements = parsed as RevocationStatement[];

  if (!verifySignatures) return statements;

  const verified: RevocationStatement[] = [];
  for (const stmt of statements) {
    const valid = await verifyRevocationSignature(stmt);
    if (valid) {
      verified.push(stmt);
    }
  }
  return verified;
}

File: packages/core-py/agentid/revocation.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

from nacl.signing import SigningKey

from agentid.identity import get_agent_id
from agentid.types import Keypair, RevocationStatement
from agentid.utils import encode_base64


def _canonicalize_revocation_payload(
    agent_id: str,
    revoked_key_id: str,
    revoked_at: str,
    reason: str,
) -> str:
    return json.dumps({
        "type": "revocation",
        "agent_id": agent_id,
        "revoked_key_id": revoked_key_id,
        "revoked_at": revoked_at,
        "reason": reason,
    }, separators=(",", ":"))


def create_revocation(
    agent_id: str,
    revoked_key_id: str,
    reason: str,
    signing_keypair: Keypair,
) -> RevocationStatement:
    revoked_at = datetime.now(timezone.utc).isoformat()
    payload = _canonicalize_revocation_payload(agent_id, revoked_key_id, revoked_at, reason)

    signing_key = SigningKey(signing_keypair.private_key)
    signed = signing_key.sign(payload.encode("utf-8"))
    signature = encode_base64(signed.signature)

    return RevocationStatement(
        type="revocation",
        agent_id=agent_id,
        revoked_key_id=revoked_key_id,
        revoked_at=revoked_at,
        reason=reason,
        signature=signature,
    )


def check_revocation(
    key_id: str,
    revocation_list: list[RevocationStatement],
) -> tuple[bool, Optional[RevocationStatement]]:
    for r in revocation_list:
        if r.revoked_key_id == key_id or r.agent_id == key_id:
            return True, r
    return False, None


def load_revocation_list(
    source: str,
    *,
    verify_signatures: bool = True,
) -> list[RevocationStatement]:
    from agentid.verification import verify_revocation_signature

    raw = Path(source).read_text(encoding="utf-8")
    data = json.loads(raw)

    if not isinstance(data, list):
        raise ValueError("Revocation list must be a JSON array")

    statements = [
        RevocationStatement(
            type=r["type"],
            agent_id=r["agent_id"],
            revoked_key_id=r["revoked_key_id"],
            revoked_at=r["revoked_at"],
            reason=r["reason"],
            signature=r["signature"],
        )
        for r in data
    ]

    if not verify_signatures:
        return statements

    return [s for s in statements if verify_revocation_signature(s)]
